### PR TITLE
Temporary solution to avoid recompilation

### PR DESCRIPTION
**Pull request by chenmoneygithub**
While we are evaluating the long term solution on how to split the mega sampler class, I am putting up this temporary fix so that we can avoid the recompilation.

### CODE REVIEW

File: keras_nlp/models/gpt2/gpt2_causal_lm.py [hunk 1]
```diff
@@ -293,10 +294,24 @@ def generate(
...
+            raise ValueError(
+                "`BeamSampler` is not supported right now, please choose "
+                "another sampler, e.g., `TopPSampler`."
+            )
+        if hasattr(self, "jit_compile"):
+            # `jit_compile` is a public property as of tf 2.12. hasattr is for
+            # backward compat.
+            sampler.jit_compile = self.jit_compile
         sampler.run_eagerly = self.run_eagerly
+        if hasattr(self, "sampler") and serialize(sampler) == serialize(
```
Comment (mattdangerw) [inline keras_nlp/models/gpt2/gpt2_causal_lm.py:307]: just do `self.sampler = None` in init. we should avoid hasattr checking with objects under our control

File: keras_nlp/samplers/sampler.py [hunk 2]
```diff
@@ -230,10 +231,12 @@ def __call__(
         # , and turn on `jit_compile` accordingly.
         sample = self.sample
         if not self.run_eagerly:
-            sample = tf.function(self.sample, jit_compile=self.jit_compile)
+            if getattr(self, "jit_compile", True):
+                sample = self._sample_graph
```
Comment (mattdangerw) [inline keras_nlp/samplers/sampler.py:235]: why not just inline all of this

`sampler = tf.function(self.sample, jit_compile=self.jit_compile)` would save a bit

also, same comment here as above. set `jit_compile = True` (or False) in init, avoid getattr

**Review by mattdangerw — CHANGES_REQUESTED** [review]
I'm down for this while I line up https://github.com/keras-team/keras-nlp/pull/804. I'll try to get that ready soon. Some small comments

File: keras_nlp/models/gpt2/gpt2_causal_lm.py (hunk 1 above)
Comment (chenmoneygithub) [reply keras_nlp/models/gpt2/gpt2_causal_lm.py:307]: sg!

File: keras_nlp/samplers/sampler.py (hunk 2 above)
Comment (chenmoneygithub) [reply keras_nlp/samplers/sampler.py:235]: `sample = tf.function(self.sample, jit_compile=self.jit_compile)` still causes recompile if the call is under a for-loop, my guess is it is considered as "defining tf.function in a for-loop". 

Fixed the getattr stuff.

File: keras_nlp/models/gpt2/gpt2_causal_lm.py [hunk 3]
```diff
@@ -293,10 +295,22 @@ def generate(
                 "`preprocessor` is set before calling `generate`."
             )
         sampler = keras_nlp.samplers.get(sampler)
-        # `jit_compile` is a property of keras.Model after tf 2.12.
-        # Use `getattr()` for backwards compatibility.
-        sampler.jit_compile = getattr(self, "jit_compile", True)
+        if sampler.__class__.__name__ == "BeamSampler":
```
Comment (mattdangerw) [inline keras_nlp/models/gpt2/gpt2_causal_lm.py:298]: Can't you just direct compare `if sampler.__class__ == BeamSampler` here? Better to do this outside string land I think

**Review by mattdangerw — APPROVED** [review]

**Pull request merged**